proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+IN{IN} -- "questions (List[str])" --> CodingLLM
+IN{IN} -- "questions (List[str])" --> MedicalLLM
+Retriever -- "documents (List[List[Doc]])"  --> LegalLLM
+Retriever -- "documents (List[List[Doc]])"  --> CodingLLM
+Retriever -- "documents (List[List[Doc]])"  --> MedicalLLM
+LegalLLM -- "replies (List[List[str]])" --> JoinLists
+CodingLLM -- "replies (List[List[str]])" --> JoinLists
+MedicalLLM -- "replies (List[List[str]])" --> JoinLists
+JoinLists -- "list (List[List[str]])" --> RepliesToAnswers
+RepliesToAnswers -- "answers (List[List[Answer]])" --> OUT{OUT}
+```
+
+Assuming the prompt requires only questions and documents, or the number of connections to make will grow. Also, each of the LLM component will perform the exact same rendering step under the hood, so it results in duplication of efforts.
+
+Compare with a pipeline that uses a `Prompts` :
+
+```mermaid
+graph TD;
+
+IN{IN} -- "questions (List[str])" --> Retriever
+IN{IN} -- "questions (List[str])" --> Prompts
+Retriever -- "documents (List[List[Doc]])"  --> Prompts
+Prompts -- "prompts (List[str])" --> LegalLLM
+Prompts -- "prompts (List[str])" --> CodingLLM
+Prompts -- "prompts (List[str])" --> MedicalLLM
+LegalLLM -- "replies (List[List[str]])" --> JoinLists
+CodingLLM -- "replies (List[List[str]])" --> JoinLists
+MedicalLLM -- "replies (List[List[str]])" --> JoinLists
+JoinLists -- "list (List[List[str]])" --> RepliesToAnswers
+RepliesToAnswers -- "answers (List[List[Answer]])" --> OUT{OUT}
+```
+
+**CON: One additional component = more complexity**
+
+The drawback is that `Prompts` is an additional component, so we must evaluate if this additional flexibility is worth the additional complexity. To do so, we need to understand how likely it is that the prompt will need to be sent to different pipelines and under which circumstances.
+
+## Output parsing
+
+LLMs clients output strings, but many components expect other object types, and LLMs may produce output in a parsable format that can be directly converted into objects. Output parsers transform these strings into objects of the user’s choosing.
+
+In Haystack 1.x, this task was assigned to the subclasses of `BaseOutputParser`. In 2.0 we’re going to have a very similar situation, with the difference that such classes are components.
+
+The most straightforward component in this category is `RepliesToAnswersConverter`. It takes the string replies of an LLM and produce `Answer` objects.  One additional output parser could be `RepliesToAnswersWithReferencesConverter`, which also connects answers to the documents used to produce them. As the need for additional output parsers arises, we will progressively add more.
+
+Draft I/O for `RepliesToAnswersConverter` (note: this may end up being almost the entire component’s implementation):
+
+```python
+@component
+class RepliesToAnswersConverter:
+
+    @component.output_types(answers=List[List[Answer]])
+    def run(self, replies: List[List[str]]):
```

Comment:
Answers have metadata, which document they are based on etc. 

Would that need to be passed to the component directly from the retriever then? Order of answers and documents would be guaranteed so that they can be matched?

Reply:
I'd say yes, that's the idea: if you want the answers to contain which documents they are generated from, connect the retriever to the converter to match them. The order is preserved so that this type of matching is possible. There should be an example about `RepliesToAnswerWithReferencesConverter` (way too long of a name :smile:) doing something similar.